proc/threads.c

```diff
@@ -2086,120 +2088,197 @@ void proc_threadsDump(u8 priority)
+	}
+
+	if ((flags & PH_THREADINFO_STATE) == PH_THREADINFO_STATE) {
+		info->state = (int)thread->state;
+	}
+
+	if ((flags & PH_THREADINFO_LOAD) == PH_THREADINFO_LOAD) {
+		if (now != thread->startTime) {
+			info->load = (int)((thread->cpuTime * 1000) / (now - thread->startTime));
 		}
 		else {
-			info[i].load = 0;
+			info->load = 0;
 		}
-		info[i].cpuTime = t->cpuTime;
+	}
 
-		if (t->state == READY && t->maxWait < now - t->readyTime) {
-			info[i].wait = now - t->readyTime;
+	if ((flags & PH_THREADINFO_WAITING) == PH_THREADINFO_WAITING) {
+		if (thread->state == READY && thread->maxWait < now - thread->readyTime) {
+			info->wait = now - thread->readyTime;
 		}
 		else {
-			info[i].wait = t->maxWait;
+			info->wait = thread->maxWait;
 		}
-		hal_spinlockClear(&threads_common.spinlock, &sc);
+	}
 
-		if (t->process != NULL) {
-			map = t->process->mapp;
-
-			if (t->process->path != NULL) {
-				space = sizeof(info[i].name);
-				name = info[i].name;
-
-				if (t->process->argv != NULL) {
-					for (argc = 0; t->process->argv[argc] != NULL; ++argc) {
-						if ((int)space <= 0) {
-							break;
-						}
-						len = min(hal_strlen(t->process->argv[argc]) + 1U, space);
-						hal_memcpy(name, t->process->argv[argc], len);
-						name[len - 1U] = ' ';
-						name += len;
-						space -= len;
-					}
-					*(name - 1) = '\0';
-				}
-				else {
-					len = hal_strlen(t->process->path) + 1U;
-					hal_memcpy(info[i].name, t->process->path, min(space, len));
-				}
+	if ((flags & PH_THREADINFO_CPUTIME) == PH_THREADINFO_CPUTIME) {
+		info->cpuTime = thread->cpuTime;
+	}
 
-				info[i].name[sizeof(info[i].name) - 1U] = '\0';
-			}
-			else {
-				info[i].name[0] = '\0';
-			}
-		}
-		else {
-			map = threads_common.kmap;
-			hal_memcpy(info[i].name, "[idle]", sizeof("[idle]"));
+	if ((flags & PH_THREADINFO_NAME) == PH_THREADINFO_NAME) {
+		_proc_makeProcessName(thread, info->name, sizeof(info->name));
+	}
+
+	if ((flags & PH_THREADINFO_VMEM) == PH_THREADINFO_VMEM) {
+		info->vmem = _proc_calculateVmem(thread);
+	}
+
+	if ((flags & PH_THREADINFO_PPID) == PH_THREADINFO_PPID) {
+		ppid = posix_getppid(info->pid);
+		if (ppid > 0) {
+			info->ppid = ppid;
 		}
+	}
+}
 
-		info[i].vmem = 0;
 
-#ifdef NOMMU
-		if (t->process != NULL) {
-			entry = t->process->entries;
-			if (entry != NULL) {
-				do {
-					info[i].vmem += (int)entry->size;
-					entry = entry->next;
-				} while (entry != t->process->entries);
-			}
+int proc_threadsList(int tid, unsigned int flags, int n, threadinfo_t *info)
+{
+	int i = 0;
+	thread_t *t;
+
+	if (info == NULL) {
+		return -EINVAL;
+	}
+
+	(void)proc_lockSet(&threads_common.lock);
+
+	if (tid == PH_THREADINFO_THREADS_ALL) {
+		t = lib_treeof(thread_t, idlinkage, lib_rbMinimum(threads_common.id.root));
+
+		if ((flags & PH_THREADINFO_OPT_THREADCOUNT) == PH_THREADINFO_OPT_THREADCOUNT) {
+			do {
+				i++;
+				t = lib_idtreeof(thread_t, idlinkage, lib_idtreeNext(&t->idlinkage.linkage));
```

Comment:
I wonder if we shouldn't just maintain a global thread count instead

Reply:
Might be, though I'd say that's good enough for the amount of changes here. We'd have to guarantee that this counter will be kept in sync with node count in tree, which I'd say is besides the point in this PR.

Reply:
Yeah, you're right. Maybe add a TODO/REVISIT then for now?